pkg/controllers/nodeclaim/podevents/controller.go

```diff
@@ -0,0 +1,128 @@
+	return &Controller{
+		kubeClient: kubeClient,
+		clock:      clk,
+	}
+}
+
+//nolint:gocyclo
+func (c *Controller) Reconcile(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	// If the pod doesn't have a node name, we don't know which node this pod refers to.
+	// or if this is a daemonset
+	if pod.Spec.NodeName == "" || podutils.IsOwnedByDaemonSet(pod) {
+		return reconcile.Result{}, nil
+	}
+
+	node := &corev1.Node{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: pod.Spec.NodeName}, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting node, %w", err)
+	}
+
+	// If the node isn't owned by Karpenter, don't do anything
+	if _, ok := node.Labels[v1.NodePoolLabelKey]; !ok {
+		return reconcile.Result{}, nil
+	}
+
+	// If there's no associated node claim, it's not a karpenter owned node.
+	nc, err := nodeutils.NodeClaimForNode(ctx, c.kubeClient, node)
+	if err != nil {
+		// if the nodeclaim doesn't exist, or has duplicates,
+		if nodeutils.IsDuplicateNodeClaimError(err) || nodeutils.IsNodeClaimNotFoundError(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("getting nodeclaims for node, %w", err)
+	}
+
+	stored := nc.DeepCopy()
+	// If we've set the lastPodEvent before and it hasn't been before the timeout, don't do anything
+	if !nc.Status.LastPodEventTime.Time.IsZero() && c.clock.Since(nc.Status.LastPodEventTime.Time) < dedupeTimeout {
+		return reconcile.Result{}, nil
+	} else {
+		// otherwise, set the pod event time to now
+		nc.Status.LastPodEventTime.Time = c.clock.Now()
+	}
+	if !equality.Semantic.DeepEqual(stored, nc) {
+		if err := c.kubeClient.Status().Patch(ctx, nc, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("nodeclaim.podevents").
+		For(&corev1.Pod{}).
+		WithEventFilter(predicate.TypedFuncs[client.Object]{
+			// If a pod is bound to a node or goes terminal
+			UpdateFunc: func(e event.TypedUpdateEvent[client.Object]) bool {
+				oldPod := (e.ObjectOld).(*corev1.Pod)
+				newPod := (e.ObjectNew).(*corev1.Pod)
+				// if this is a newly bound pod
+				bound := oldPod.Spec.NodeName == "" && newPod.Spec.NodeName != ""
+				// if this is a newly terminal pod
+				terminal := (newPod.Spec.NodeName != "" && !podutils.IsTerminal(oldPod) && podutils.IsTerminal(newPod))
+				// if this is a newly terminating pod
+				terminating := (newPod.Spec.NodeName != "" && !podutils.IsTerminating(oldPod) && podutils.IsTerminating(newPod))
+				// return true if it was bound to a node, went terminal, or went terminating
+				return bound || terminal || terminating
+			},
+			// If the pod is deleted, but wasn't already treated as terminal or terminating
+			DeleteFunc: func(e event.TypedDeleteEvent[client.Object]) bool {
```

Comment:
Didn't we decide to drop this section since we agreed its immaterial to this controller?